docs/proposals/edge-node-tasks-design.md

```diff
@@ -0,0 +1,306 @@
+type NodeUpgradeJobList struct {
+	// Standard type metadata.
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard list metadata.
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// List of NodeUpgradeJobs.
+	Items []NodeUpgradeJob `json:"items"`
+}
+
+``// NodeUpgradeJobSpec is the specification of the desired behavior of the NodeUpgradeJob.
+// NodeUpgradeJobSpec is the specification of the desired behavior of the NodeUpgradeJob.
+type NodeUpgradeJobSpec struct {
+    // +Required: Version is the EdgeCore version to upgrade.
+    Version string `json:"version,omitempty"`
+    // TimeoutSeconds limits the duration of the node upgrade job.
+    // Default to 300.
+    // If set to 0, we'll use the default value 300.
+    // +optional
+    TimeoutSeconds *uint32 `json:"timeoutSeconds,omitempty"`
+    // NodeNames is a request to select some specific nodes. If it is non-empty,
+    // the upgrade job simply select these edge nodes to do upgrade operation.
+    // Please note that sets of NodeNames and LabelSelector are ORed.
+    // Users must set one and can only set one.
+    // +optional
+    NodeNames []string `json:"nodeNames,omitempty"`
+    // LabelSelector is a filter to select member clusters by labels.
+    // It must match a node's labels for the NodeUpgradeJob to be operated on that node.
+    // Please note that sets of NodeNames and LabelSelector are ORed.
+    // Users must set one and can only set one.
+    // +optional
+    LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+    // Image specifies a container image name, the image contains: keadm and edgecore.
+    // keadm is used as upgradetool, to install the new version of edgecore.
+    // The image name consists of registry hostname and repository name,
+    // if it includes the tag or digest, the tag or digest will be overwritten by Version field above.
+    // If the registry hostname is empty, docker.io will be used as default.
+    // The default image name is: kubeedge/installation-package.
+    // +optional
+    Image string `json:"image,omitempty"`
+    // Concurrency specifies the max number of edge nodes that can be upgraded at the same time.
+    // The default Concurrency value is 1.
+    // +optional
+    Concurrency int32 `json:"concurrency,omitempty"`
+    
+    // CheckItems specifies the items need to be checked before the task is executed.
+    // The default CheckItems value is nil.
+    // +optional
+    CheckItems []string `json:"checkItems,omitempty"`
+
+    FailureTolerate float64 `json:"failureTolerate,omitempty"`
+}
+
+// NodeUpgradeJobStatus stores the status of NodeUpgradeJob.
+// contains multiple edge nodes upgrade status.
+// +kubebuilder:validation:Type=object
+type NodeUpgradeJobStatus struct {
+	// State represents for the state phase of the NodeUpgradeJob.
+	// There are three possible state values: "", upgrading and completed.
+	State UpgradeState `json:"state,omitempty"`
+    currentVersion string
```

Comment:
On second thought, I found it doesn't make sense to put the CurrentVersion and HistoricVersion in this structure. This is assumed that the base version of the node associated with the node upgrade task is the same. But in reality, several nodes of different older versions will be upgraded to the latest version.

Reply:
Yes. Currently, it appears that if the upgrade procedure is only A -> C, you may simply use State+Event to determine whether it was successful. However, the idea of multi-version upgrades can be incorporated later(A -> B -> C). Also, exercise judgment before updating to confirm that the node's historical version is consistent.